### ISSUE TITLE
502 error - Authentication popup never shown in browser

### ISSUE THREAD

**Issue by srinisubramanian** [dutchcoders/transfer.sh#305]
I have a private transfer.sh instance running with authentication switched on.  The setup is as follows:

1.  transfer.sh running on http://localhost:4000 on Centos 7
2.  Exposed to the internet via nginx with http redirect to https
3. X-Forwarded-Proto  is set so that transfer.sh displays all urls as https

Issue:

1.  When I upload small files of a few KB the browser pop up for auth appears and accepts the user id and password.
2.  When I upload larger files (of a few MB) that take a few seconds to upload, the authentication popup never appears and after a few seconds after reaching 100% a (502) error is displayed like this: Error (502) during upload of file  .....
3.  In the transfer.sh server log a 401 error is recorded
4.  The workaround is to first upload a small file, authenticate and then upload the large file and it then works.

Any ideas if this is a config/landscape fix or needs a code change?

**Comment by aspacca** [comment]
@srinisubramanian I have exactly the same setuo and I couldn't reproduce the issue

can you post your configuration for nginx and the args for transfer.sh?
also, can you test uploading the file directly to transfer.sh with some client that support the authentication challenge?

it seems to be a issue on the nginx side

**Comment by srinisubramanian** [comment]
In your case does the auth pop up immediately on the web page or after the upload is complete?

I will post the config shortly.

**Comment by aspacca** [comment]
after the progress bar in the frontend is complete

I was able to reproduce the issue letting transfer.sh listen on http and use the `X-Forwarded-Proto` header

in my current setup transer.sh is also listening on https
```
  location / {
    proxy_set_header Host $host;
    proxy_set_header X-Forwarded-For $remote_addr;
    proxy_pass https://127.0.0.1:4000/;
  }
```

**Comment by srinisubramanian** [comment]
Yes. My setup runs on http. If you can find an issue let me know. Else maybe I need to run internal on a self signed https certificate.

**Comment by aspacca** [comment]
if the proxy is on http it has a certificate, so you can use the same on transfer.sh (that's what I do)
I will investigate a little more on the problem

**Comment by srinisubramanian** [comment]
Ok. So you are suggesting I just turn on https? Will do and check.

**Comment by aspacca** [comment]
the problem seems to be on how nginx expects response from the server when uploading file.
the transfer.sh upstream is sending a 401 before nginx finishes to upload the file, causing the 502

now I have the same issue even with https upstream

**Comment by aspacca** [comment]
https://blog.bigbinary.com/2016/01/07/not-using-https-might-be-breaking-file-uploads.html

**Comment by aspacca** [comment]
it should be solved with http upstream and `X-Forwarded-Proto` and setting the following in the block with the proxy directive:
```
    keepalive_requests 0;
    keepalive_timeout 0s;
```
can you please try?

**Comment by srinisubramanian** [comment]
No luck @aspacca Not working.  Still returns the 502 error

Here is my nginx config.

        location / {
                proxy_set_header X-Real-IP  $remote_addr;
                proxy_set_header X-Forwarded-For $remote_addr;
                proxy_set_header X-Forwarded-Proto $scheme;
                proxy_set_header Host $host;
                keepalive_requests 0;
                keepalive_timeout 0s;
                proxy_pass http://127.0.0.1:9400;
                client_max_body_size 100m;
                proxy_read_timeout 600s;
        }

**Comment by aspacca** [comment]
damn, I tested only once, it worked, now not anymore
can you ask in the nginx mailing list?

**Comment by srinisubramanian** [comment]
Will do

**Comment by aspacca** [comment]
thanks

**Comment by srinisubramanian** [comment]
Here’s what I did for now.

1. Switched off auth in transfer.sh
2. Enabled http basic auth in nginx for the transfersh site except for GET
3. This allows site to load without auth, download to happen without auth and upload prompts for password immediately.

I can live with this for now :)

**Comment by srinisubramanian** [comment]
Here’s what I did for now.

- Switched off auth in transfer.sh  
- Enabled http basic auth in nginx for the transfersh site except for GET  
- This allows site to load without auth, download to happen without auth and upload prompts for password immediately  

I can live with this for now :)

**Comment by aspacca** [comment]
just to know, did you asked in the nginx mailing list?
otherwise I will do

**Comment by srinisubramanian** [comment]
I will ask tomorrow.  If you want to go ahead

**Comment by aspacca** [comment]
oki thanks

**Comment by srinisubramanian** [comment]
Having trouble getting on to that mailing list. Confirmation mail never reaches, tried a couple. 

@aspacca can you please post?

**Comment by aspacca** [comment]
@srinisubramanian just subcribed to the mailing list and waiting for confirmation email as well

https://forum.nginx.org may be an alternative if you want to try (https://forum.nginx.org/read.php?2,255)

**Comment by aspacca** [comment]
@srinisubramanian I found the solution:
`proxy_request_buffering off;`

available from nginx 1.7.11: https://nginx.org/en/docs/http/ngx_http_proxy_module.html#proxy_request_buffering

**Comment by srinisubramanian** [comment]
@aspacca This works, thanks.  Also the behavior is such that the user id/pwd is prompted right at the beginning which is good.  Thanks for following up on this.
